channeldb/invoices.go

```diff
@@ -546,28 +546,42 @@ func (d *DB) LookupInvoice(paymentHash [32]byte) (Invoice, error) {
 	return invoice, nil
 }
 
+// InvoiceWithPaymentHash is used to store an invoice and its corresponding
+// payment hash. This struct is only used to store results of
+// ChannelDB.FetchAllInvoicesWithPaymentHash() call.
+type InvoiceWithPaymentHash struct {
+	Invoice     Invoice
+	PaymentHash lntypes.Hash
+}
+
 // FetchAllInvoicesWithPaymentHash returns all invoices and their payment hashes
 // currently stored within the database. If the pendingOnly param is true, then
 // only unsettled invoices and their payment hashes will be returned, skipping
-// all invoices that are fully settled.
+// all invoices that are fully settled or canceled. Note that the returned
+// array is not ordered by add index.
 func (d *DB) FetchAllInvoicesWithPaymentHash(pendingOnly bool) (
-	[]Invoice, []lntypes.Hash, error) {
+	[]InvoiceWithPaymentHash, error) {
```

Comment:
i wonder, is there reason this can't be done when FetchAllInvoicesWithPaymentHash is introduced? is there another dependency that prevents that?

Reply:
Do you mean why we don't just return an optional `PaymentHash` field in the `Invoice`? I had an incarnation (lost in the numerous reviews) where this was done so, but seems like the introduced inconsistency that not all calls can populate the payment hash due to how the indexes are done wasn't worth it and reviewers requested to have a separate `FetchAllInvoicesWithPaymentHash` call instead. 

The reason why this is hard to do for all calls is because we have to iterate on the `invoiceIndex` as that maps from payment hash => invoice add idx, while the `invoiceBucket` maps from invoice add idx => invoice. So only calls where we refer to invoices by their payment hash can be changed to populate payment hash.